File: server/src/main/java/com/netmarble/chat/application/service/ChatRoomApplicationService.java
package com.netmarble.chat.application.service;

import com.netmarble.chat.application.dto.*;
import com.netmarble.chat.domain.model.ChatRoom;
import com.netmarble.chat.domain.model.ChatRoomMember;
import com.netmarble.chat.domain.model.Message;
import com.netmarble.chat.domain.model.User;
import com.netmarble.chat.domain.repository.ChatRoomMemberRepository;
import com.netmarble.chat.domain.repository.ChatRoomRepository;
import com.netmarble.chat.domain.repository.MessageRepository;
import com.netmarble.chat.domain.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.messaging.simp.SimpMessagingTemplate;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.Comparator;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;

/**
 * ChatRoom 애플리케이션 서비스
 */
@Slf4j
@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class ChatRoomApplicationService {

    private final ChatRoomRepository chatRoomRepository;
    private final UserRepository userRepository;
    private final MessageRepository messageRepository;
    private final ChatRoomMemberRepository chatRoomMemberRepository;
    private final ReadStatusApplicationService readStatusApplicationService;
    private final SimpMessagingTemplate messagingTemplate;

    /**
     * 채팅방 생성
     */
    @Transactional
    public ChatRoomResponse createChatRoom(CreateChatRoomRequest request) {
        log.info("Creating chat room: {}", request.getName());
        
        User creator = userRepository.findById(request.getCreatorId())
            .orElseThrow(() -> new IllegalArgumentException("사용자를 찾을 수 없습니다: " + request.getCreatorId()));
        
        ChatRoom chatRoom = new ChatRoom(request.getName(), request.getImageUrl(), creator);
        ChatRoom savedChatRoom = chatRoomRepository.save(chatRoom);
        
        // 시스템 메시지: 채팅방 생성
        Message systemMessage = Message.createSystemMessage(
            savedChatRoom, 
            creator.getNickname() + "님이 채팅방을 생성했습니다."
        );
        messageRepository.save(systemMessage);
        
        log.info("Chat room created: id={}, name={}", savedChatRoom.getId(), savedChatRoom.getName());
        
        return ChatRoomResponse.from(savedChatRoom);
    }

    /**
     * 채팅방 목록 조회 (활성 채팅방, isMember/읽지 않은 수/마지막 메시지/멤버 아바타 포함)
     */
    public List<ChatRoomResponse> getAllActiveChatRooms(Long userId) {
        // 최근 100개 방만 조회 (N+1 쿼리 부하 제한 - 방이 누적될수록 응답 지연 방지)
        List<ChatRoom> chatRooms = chatRoomRepository.findAllActive().stream()
            .limit(100)
            .collect(Collectors.toList());

        // 단일 쿼리로 유저가 참가 중인 채팅방 ID Set 조회 (lazy loading 의존 제거)
        Set<Long> memberRoomIds = (userId != null)
            ? chatRoomMemberRepository.findActiveChatRoomIdsByUserId(userId)
            : Set.of();

        return chatRooms.stream()
            .map(chatRoom -> buildRoomResponse(chatRoom, userId, memberRoomIds))
            .collect(Collectors.toList());
    }

    /**
     * 채팅방 응답 DTO 빌드 (공통 로직)
     */
    private ChatRoomResponse buildRoomResponse(ChatRoom chatRoom, Long userId, Set<Long> memberRoomIds) {
        // 1. 멤버 여부 — DB에서 직접 조회한 Set으로 판단 (lazy loading 불필요)
        boolean isMember = memberRoomIds.contains(chatRoom.getId());

        // 2. 읽지 않은 메시지 수 — 참가 중인 방에만 계산
        long unreadCount = (userId != null && isMember)
            ? readStatusApplicationService.getUnreadCount(userId, chatRoom.getId())
            : 0L;

        // 3. 마지막 메시지 — 참가 중인 방에만 조회 (비멤버는 내용 미공개)
        String lastMessageContent = null;
        LocalDateTime lastMessageAt = null;
        if (isMember) {
            Optional<Message> lastMsgOpt = messageRepository.findLastByChatRoomId(chatRoom.getId());
            lastMessageContent = lastMsgOpt.map(m -> switch (m.getType()) {
                case IMAGE   -> "[사진]";
                case STICKER -> "[스티커]";
                default      -> m.getContent();
            }).orElse(null);
            lastMessageAt = lastMsgOpt.map(Message::getSentAt).orElse(null);
        }

        // 4. 멤버 아바타 (최대 4명 — 현재 참여 중인 멤버만, 본인 항상 제외, 입장 시간순)
        List<ChatRoomResponse.MemberAvatar> memberAvatars = chatRoom.getMembers().stream()
            .filter(ChatRoomMember::isActive)
            .filter(m -> userId == null || !m.getUser().getId().equals(userId))
            .sorted(Comparator.comparing(ChatRoomMember::getJoinedAt))
            .limit(4)
            .map(m -> new ChatRoomResponse.MemberAvatar(
                m.getUser().getProfileColor(),
                m.getUser().getProfileImage(),
                m.getUser().getNickname()
            ))
            .collect(Collectors.toList());

        return ChatRoomResponse.builder()
            .id(chatRoom.getId())
            .name(chatRoom.getName())
            .imageUrl(chatRoom.getImageUrl())
            .creatorId(chatRoom.getCreator().getId())
            .creatorNickname(chatRoom.getCreator().getNickname())
            .createdAt(chatRoom.getCreatedAt())
            .active(chatRoom.isActive())
            .memberCount(chatRoom.getActiveMemberCount())
            .unreadCount(unreadCount)
            .isMember(isMember)
            .lastMessageContent(lastMessageContent)
            .lastMessageAt(lastMessageAt)
            .memberAvatars(memberAvatars)
            .build();
    }

    /**
     * 채팅방 상세 조회
     */
    public ChatRoomResponse getChatRoomById(Long id) {
        ChatRoom chatRoom = chatRoomRepository.findById(id)
            .orElseThrow(() -> new IllegalArgumentException("채팅방을 찾을 수 없습니다: " + id));
        return ChatRoomResponse.from(chatRoom);
    }

    /**
     * 채팅방 입장
     */
    @Transactional
    public ChatRoomResponse joinChatRoom(JoinChatRoomRequest request) {
        log.info("User {} joining chat room {}", request.getUserId(), request.getChatRoomId());
        
        ChatRoom chatRoom = chatRoomRepository.findById(request.getChatRoomId())
            .orElseThrow(() -> new IllegalArgumentException("채팅방을 찾을 수 없습니다: " + request.getChatRoomId()));
        
        User user = userRepository.findById(request.getUserId())
            .orElseThrow(() -> new IllegalArgumentException("사용자를 찾을 수 없습니다: " + request.getUserId()));
        
        // 멤버 추가 (중복 입장 시 false 반환)
        boolean isNewJoin = chatRoom.addMember(user);

        if (isNewJoin) {
            // 새로 입장하거나 재입장한 경우에만 시스템 메시지 생성
            Message systemMessage = Message.createSystemMessage(
                chatRoom,
                user.getNickname() + "님이 입장했습니다."
            );
            Message savedSystemMessage = messageRepository.save(systemMessage);

            // WebSocket을 통해 실시간으로 브로드캐스트
            MessageResponse response = MessageResponse.from(savedSystemMessage);
            messagingTemplate.convertAndSend(
                "/topic/chatroom/" + chatRoom.getId(),
                response
            );

            log.info("User {} joined chat room {} (new join)", user.getNickname(), chatRoom.getName());
        } else {
            // 이미 활성 멤버인 경우 - 읽음 처리만 수행
            log.info("User {} already in chat room {} (marking as read)", user.getNickname(), chatRoom.getName());
        }

        // 입장 시 마지막 메시지까지 읽음 처리 (최신 메시지 1건만 조회)
        messageRepository.findLastByChatRoomId(chatRoom.getId())
            .ifPresent(lastMessage -> {
                chatRoom.getMembers().stream()
                    .filter(member -> member.getUser().getId().equals(user.getId()))
                    .findFirst()
                    .ifPresent(member -> member.updateLastReadMessage(lastMessage));
                log.info("Last read message updated for user {} in chat room {}", user.getNickname(), chatRoom.getName());
            });

        chatRoomRepository.save(chatRoom);
        
        return ChatRoomResponse.from(chatRoom);
    }

    /**
     * 채팅방 퇴장
     */
    @Transactional
    public void leaveChatRoom(Long chatRoomId, Long userId) {
        log.info("User {} leaving chat room {}", userId, chatRoomId);
        
        ChatRoom chatRoom = chatRoomRepository.findById(chatRoomId)
            .orElseThrow(() -> new IllegalArgumentException("채팅방을 찾을 수 없습니다: " + chatRoomId));
        
        User user = userRepository.findById(userId)
            .orElseThrow(() -> new IllegalArgumentException("사용자를 찾을 수 없습니다: " + userId));
        
        chatRoom.removeMember(user);
        chatRoomRepository.save(chatRoom);
        
        // 시스템 메시지: 사용자 퇴장
        Message systemMessage = Message.createSystemMessage(
            chatRoom,
            user.getNickname() + "님이 퇴장했습니다."
        );
        Message savedSystemMessage = messageRepository.save(systemMessage);
        
        // WebSocket을 통해 실시간으로 브로드캐스트
        MessageResponse response = MessageResponse.from(savedSystemMessage);
        messagingTemplate.convertAndSend(
            "/topic/chatroom/" + chatRoom.getId(),
            response
        );
        
        log.info("User {} left chat room {}", user.getNickname(), chatRoom.getName());
    }

    /**
     * 채팅방 멤버 목록 조회 (활성 멤버만)
     */
    public List<ChatRoomMemberResponse> getActiveChatRoomMembers(Long chatRoomId) {
        ChatRoom chatRoom = chatRoomRepository.findById(chatRoomId)
            .orElseThrow(() -> new IllegalArgumentException("채팅방을 찾을 수 없습니다: " + chatRoomId));
        
        return chatRoom.getMembers().stream()
            .filter(member -> member.isActive())
            .map(ChatRoomMemberResponse::from)
            .collect(Collectors.toList());
    }
    
    /**
     * 채팅방 멤버 활성 상태 업데이트
     * 온라인 상태로 전환 시 자동으로 읽음 처리 수행
     */
    @Transactional
    public void updateMemberActiveStatus(Long chatRoomId, Long userId, boolean online) {
        log.info("Updating member active status: chatRoomId={}, userId={}, online={}", 
                 chatRoomId, userId, online);
        
        ChatRoom chatRoom = chatRoomRepository.findById(chatRoomId)
            .orElseThrow(() -> new IllegalArgumentException("채팅방을 찾을 수 없습니다: " + chatRoomId));
        
        chatRoom.getMembers().stream()
            .filter(member -> member.getUser().getId().equals(userId))
            .findFirst()
            .ifPresent(member -> {
                member.updateActiveStatus(online);
                
                // 온라인 상태로 전환 시 마지막 메시지까지 읽음 처리 (최신 메시지 1건만 조회)
                if (online) {
                    messageRepository.findLastByChatRoomId(chatRoomId)
                        .ifPresent(lastMessage -> {
                            member.updateLastReadMessage(lastMessage);
                            log.info("Auto-marked as read for user {} when going online in chat room {}", userId, chatRoomId);
                        });
                }
            });
        
        chatRoomRepository.save(chatRoom);
    }
    
    /**
     * 채팅방 멤버 활동 업데이트 (하트비트)
     * 활동 업데이트 시 자동으로 읽음 처리도 수행
     */
    @Transactional
    public void updateMemberActivity(Long chatRoomId, Long userId) {
        ChatRoom chatRoom = chatRoomRepository.findById(chatRoomId)
            .orElseThrow(() -> new IllegalArgumentException("채팅방을 찾을 수 없습니다: " + chatRoomId));
        
        chatRoom.getMembers().stream()
            .filter(member -> member.getUser().getId().equals(userId))
            .findFirst()
            .ifPresent(member -> {
                member.updateActivity();
                // 활동 중인 사용자는 자동으로 읽음 처리 (최신 메시지 1건만 조회)
                messageRepository.findLastByChatRoomId(chatRoomId)
                    .ifPresent(member::updateLastReadMessage);
            });
        
        chatRoomRepository.save(chatRoom);
    }
}

File: chat-server/src/application/service/chat-room-application.service.ts
import { Inject, Injectable, Logger } from '@nestjs/common';
import { ChatRoom } from '../../domain/model/chat-room.js';
import { ChatRoomMember } from '../../domain/model/chat-room-member.js';
import { Message, MessageType } from '../../domain/model/message.js';
import {
  USER_REPOSITORY,
  UserRepository,
} from '../../domain/repository/user.repository.js';
import {
  CHAT_ROOM_REPOSITORY,
  ChatRoomRepository,
} from '../../domain/repository/chat-room.repository.js';
import {
  MESSAGE_REPOSITORY,
  MessageRepository,
} from '../../domain/repository/message.repository.js';
import {
  CHAT_ROOM_MEMBER_REPOSITORY,
  ChatRoomMemberRepository,
} from '../../domain/repository/chat-room-member.repository.js';
import { StompBrokerService } from '../../infrastructure/config/stomp-broker.service.js';
import {
  ChatRoomResponse,
  ChatRoomMemberResponse,
  CreateChatRoomRequest,
  JoinChatRoomRequest,
  MemberAvatar,
  MessageResponse,
} from '../dto/index.js';

@Injectable()
export class ChatRoomApplicationService {
  private readonly logger = new Logger(ChatRoomApplicationService.name);

  constructor(
    @Inject(CHAT_ROOM_REPOSITORY)
    private readonly chatRoomRepository: ChatRoomRepository,
    @Inject(USER_REPOSITORY)
    private readonly userRepository: UserRepository,
    @Inject(MESSAGE_REPOSITORY)
    private readonly messageRepository: MessageRepository,
    @Inject(CHAT_ROOM_MEMBER_REPOSITORY)
    private readonly chatRoomMemberRepository: ChatRoomMemberRepository,
    private readonly stompBroker: StompBrokerService,
  ) {}

  async createChatRoom(
    request: CreateChatRoomRequest,
  ): Promise<ChatRoomResponse> {
    const creator = await this.userRepository.findById(request.creatorId);
    if (!creator)
      throw new Error('사용자를 찾을 수 없습니다: ' + request.creatorId);

    const chatRoom = new ChatRoom(
      request.name,
      request.imageUrl ?? null,
      creator.id!,
    );
    const saved = await this.chatRoomRepository.save(chatRoom);

    // 생성자를 멤버로 추가
    const member = new ChatRoomMember(saved.id!, creator.id!);
    await this.chatRoomMemberRepository.save(member);

    // 시스템 메시지
    const sysMsg = Message.createSystemMessage(
      saved.id!,
      `${creator.nickname}님이 채팅방을 생성했습니다.`,
    );
    await this.messageRepository.save(sysMsg);

    return this.buildBasicResponse(saved, creator.nickname);
  }

  async getAllActiveChatRooms(
    userId: number | null,
  ): Promise<ChatRoomResponse[]> {
    const chatRooms = (await this.chatRoomRepository.findAllActive()).slice(
      0,
      100,
    );

    const memberRoomIds = userId
      ? await this.chatRoomMemberRepository.findActiveChatRoomIdsByUserId(
          userId,
        )
      : new Set<number>();

    const results: ChatRoomResponse[] = [];
    for (const room of chatRooms) {
      results.push(
        await this.buildRoomResponse(room, userId, memberRoomIds),
      );
    }
    return results;
  }

  async getChatRoomById(id: number): Promise<ChatRoomResponse> {
    const room = await this.chatRoomRepository.findById(id);
    if (!room) throw new Error('채팅방을 찾을 수 없습니다: ' + id);
    const creator = await this.userRepository.findById(room.creatorId);
    return this.buildBasicResponse(room, creator?.nickname ?? 'Unknown');
  }

  async joinChatRoom(request: JoinChatRoomRequest): Promise<ChatRoomResponse> {
    const chatRoom = await this.chatRoomRepository.findById(
      request.chatRoomId,
    );
    if (!chatRoom)
      throw new Error('채팅방을 찾을 수 없습니다: ' + request.chatRoomId);

    const user = await this.userRepository.findById(request.userId);
    if (!user) throw new Error('사용자를 찾을 수 없습니다: ' + request.userId);

    // 기존 활성 멤버 확인
    let member =
      await this.chatRoomMemberRepository.findActiveByChatRoomIdAndUserId(
        request.chatRoomId,
        request.userId,
      );

    let isNewJoin = false;
    if (!member) {
      // 비활성 멤버 확인 (leave 후 재입장)
      const existing =
        await this.chatRoomMemberRepository.findByChatRoomIdAndUserId(
          request.chatRoomId,
          request.userId,
        );

      if (existing && !existing.active) {
        existing.rejoin();
        member = await this.chatRoomMemberRepository.save(existing);
      } else {
        member = await this.chatRoomMemberRepository.save(
          new ChatRoomMember(request.chatRoomId, request.userId),
        );
      }
      isNewJoin = true;
    }

    if (isNewJoin) {
      const sysMsg = Message.createSystemMessage(
        chatRoom.id!,
        `${user.nickname}님이 입장했습니다.`,
      );
      const savedMsg = await this.messageRepository.save(sysMsg);
      this.broadcastMessage(chatRoom.id!, savedMsg, null);
    }

    // 마지막 메시지까지 읽음 처리
    const lastMsg = await this.messageRepository.findLastByChatRoomId(
      chatRoom.id!,
    );
    if (lastMsg) {
      member.updateLastReadMessage(lastMsg.id!);
      member = await this.chatRoomMemberRepository.save(member);
    }

    const creator = await this.userRepository.findById(chatRoom.creatorId);
    return this.buildBasicResponse(chatRoom, creator?.nickname ?? 'Unknown');
  }

  async leaveChatRoom(chatRoomId: number, userId: number): Promise<void> {
    const chatRoom = await this.chatRoomRepository.findById(chatRoomId);
    if (!chatRoom) throw new Error('채팅방을 찾을 수 없습니다: ' + chatRoomId);

    const user = await this.userRepository.findById(userId);
    if (!user) throw new Error('사용자를 찾을 수 없습니다: ' + userId);

    const member =
      await this.chatRoomMemberRepository.findActiveByChatRoomIdAndUserId(
        chatRoomId,
        userId,
      );
    if (!member) throw new Error('채팅방에 참여하지 않은 사용자입니다.');
    member.leave();
    await this.chatRoomMemberRepository.save(member);

    const sysMsg = Message.createSystemMessage(
      chatRoomId,
      `${user.nickname}님이 퇴장했습니다.`,
    );
    const savedMsg = await this.messageRepository.save(sysMsg);
    this.broadcastMessage(chatRoomId, savedMsg, null);
  }

  async getActiveChatRoomMembers(
    chatRoomId: number,
  ): Promise<ChatRoomMemberResponse[]> {
    const members =
      await this.chatRoomMemberRepository.findActiveByChatRoomId(chatRoomId);
    const results: ChatRoomMemberResponse[] = [];
    for (const m of members) {
      const user = await this.userRepository.findById(m.userId);
      if (!user) continue;
      const dto = new ChatRoomMemberResponse();
      dto.id = m.id!;
      dto.userId = user.id!;
      dto.nickname = user.nickname;
      dto.profileColor = user.profileColor;
      dto.profileImage = user.profileImage;
      dto.joinedAt = m.joinedAt;
      dto.leftAt = m.leftAt;
      dto.active = m.active;
      dto.lastReadMessageId = m.lastReadMessageId;
      results.push(dto);
    }
    return results;
  }

  async updateMemberActiveStatus(
    chatRoomId: number,
    userId: number,
    online: boolean,
  ): Promise<void> {
    const member =
      await this.chatRoomMemberRepository.findActiveByChatRoomIdAndUserId(
        chatRoomId,
        userId,
      );
    if (!member) return;
    member.updateActiveStatus(online);
    if (online) {
      const lastMsg =
        await this.messageRepository.findLastByChatRoomId(chatRoomId);
      if (lastMsg) member.updateLastReadMessage(lastMsg.id!);
    }
    await this.chatRoomMemberRepository.save(member);
  }

  async updateMemberActivity(
    chatRoomId: number,
    userId: number,
  ): Promise<void> {
    const member =
      await this.chatRoomMemberRepository.findActiveByChatRoomIdAndUserId(
        chatRoomId,
        userId,
      );
    if (!member) return;
    member.updateActivity();
    const lastMsg =
      await this.messageRepository.findLastByChatRoomId(chatRoomId);
    if (lastMsg) member.updateLastReadMessage(lastMsg.id!);
    await this.chatRoomMemberRepository.save(member);
  }

  private broadcastMessage(
    chatRoomId: number,
    message: Message,
    senderNickname: string | null,
  ): void {
    const response = new MessageResponse();
    response.id = message.id ?? null;
    response.chatRoomId = chatRoomId;
    response.senderId = message.sender?.id ?? null;
    response.senderNickname = senderNickname ?? 'System';
    response.content = message.content;
    response.type = message.type;
    response.messageType = message.type;
    response.sentAt = message.sentAt;
    response.deleted = message.deleted;
    this.stompBroker.send(
      `/topic/chatroom.${chatRoomId}`,
      JSON.stringify(response),
    );
  }

  private async buildRoomResponse(
    room: ChatRoom,
    userId: number | null,
    memberRoomIds: Set<number>,
  ): Promise<ChatRoomResponse> {
    const isMember = memberRoomIds.has(room.id!);
    const creator = await this.userRepository.findById(room.creatorId);
    const members = await this.chatRoomMemberRepository.findActiveByChatRoomId(
      room.id!,
    );

    let lastMessageContent: string | null = null;
    let lastMessageAt: Date | null = null;
    let unreadCount = 0;

    if (isMember) {
      const lastMsg = await this.messageRepository.findLastByChatRoomId(
        room.id!,
      );
      if (lastMsg) {
        lastMessageContent =
          lastMsg.type === MessageType.IMAGE
            ? '[사진]'
            : lastMsg.type === MessageType.STICKER
              ? '[스티커]'
              : lastMsg.content;
        lastMessageAt = lastMsg.sentAt;
      }

      if (userId) {
        const member =
          await this.chatRoomMemberRepository.findActiveByChatRoomIdAndUserId(
            room.id!,
            userId,
          );
        if (member) {
          const allMsgs =
            await this.messageRepository.findByChatRoomIdOrderBySentAtAsc(
              room.id!,
            );
          unreadCount = this.calculateUnreadForMember(
            allMsgs,
            member,
            userId,
          );
        }
      }
    }

    // 멤버 아바타 (최대 4명, 본인 제외)
    const avatarMembers = members
      .filter((m) => userId === null || m.userId !== userId)
      .sort((a, b) => a.joinedAt.getTime() - b.joinedAt.getTime())
      .slice(0, 4);

    const avatars: MemberAvatar[] = [];
    for (const m of avatarMembers) {
      const u = await this.userRepository.findById(m.userId);
      if (u) {
        const avatar = new MemberAvatar();
        avatar.profileColor = u.profileColor;
        avatar.profileImage = u.profileImage;
        avatar.nickname = u.nickname;
        avatars.push(avatar);
      }
    }

    const dto = new ChatRoomResponse();
    dto.id = room.id!;
    dto.name = room.name;
    dto.imageUrl = room.imageUrl;
    dto.creatorId = room.creatorId;
    dto.creatorNickname = creator?.nickname ?? 'Unknown';
    dto.createdAt = room.createdAt;
    dto.active = room.active;
    dto.memberCount = members.length;
    dto.unreadCount = unreadCount;
    dto.isMember = isMember;
    dto.lastMessageContent = lastMessageContent;
    dto.lastMessageAt = lastMessageAt;
    dto.memberAvatars = avatars;
    return dto;
  }

  private calculateUnreadForMember(
    allMsgs: Message[],
    member: ChatRoomMember,
    userId: number,
  ): number {
    if (!member.lastReadMessageId) {
      return allMsgs.filter(
        (m) => m.sender && m.sender.id !== userId,
      ).length;
    }
    let foundLastRead = false;
    let count = 0;
    for (const msg of allMsgs) {
      if (msg.id === member.lastReadMessageId) {
        foundLastRead = true;
        continue;
      }
      if (foundLastRead && msg.sender && msg.sender.id !== userId) {
        count++;
      }
    }
    return count;
  }

  private buildBasicResponse(
    room: ChatRoom,
    creatorNickname: string,
  ): ChatRoomResponse {
    const dto = new ChatRoomResponse();
    dto.id = room.id!;
    dto.name = room.name;
    dto.imageUrl = room.imageUrl;
    dto.creatorId = room.creatorId;
    dto.creatorNickname = creatorNickname;
    dto.createdAt = room.createdAt;
    dto.active = room.active;
    dto.memberCount = 0;
    return dto;
  }
}
